File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/seleccionable/AbstractListaSeleccionable.java
package com.victorlh.android.framework.listas.adapter.seleccionable;

import androidx.annotation.NonNull;

import com.victorlh.android.framework.listas.ItemLista;
import com.victorlh.android.framework.listas.adapter.AbstractLista;

import java.util.List;

/**
 * Created by victor on 07/07/2017.
 */
public abstract class AbstractListaSeleccionable<T extends ItemLista> extends AbstractLista<T> implements ListaSeleccionable<T> {

	private ListaSeleccionableAdapter<T> adapter;

	protected AbstractListaSeleccionable(Class<T> clazz) {
		this(clazz, ETipoSeleccion.NONE);
	}

	protected AbstractListaSeleccionable(Class<T> clazz, ETipoSeleccion tipoSeleccion) {
		super(clazz);
		this.adapter = new ListaSeleccionableAdapter<>(tipoSeleccion);
		this.adapter.attachLista(this);
	}

	@Override
	public void setTipoSeleccion(@NonNull ETipoSeleccion tipoSeleccion) {
		adapter.setTipoSeleccion(tipoSeleccion);
	}

	@Override
	public ETipoSeleccion getTipoSeleccion() {
		return adapter.getTipoSeleccion();
	}

	@Override
	public void setSeleccion(@NonNull T[] items) {
		adapter.setSeleccion(items);
	}

	@Override
	public void setSeleccion(@NonNull List<T> items) {
		adapter.setSeleccion(items);
	}

	@Override
	public void addSeleccion(@NonNull T item) {
		adapter.addSeleccion(item);
	}

	@Override
	public void removeSeleccion(@NonNull T item) {
		adapter.removeSeleccion(item);
	}

	@Override
	public void removeAllSeleccion() {
		adapter.removeAllSeleccion();
	}

	@Override
	public boolean haySeleccionados() {
		return adapter.haySeleccionados();
	}

	@Override
	public boolean isSeleccionado(@NonNull T item) {
		return adapter.isSeleccionado(item);
	}

	@Override
	public T[] getItemsSeleccionados() {
		return adapter.getItemsSeleccionados();
	}

	@Override
	public void addOnStartSeleccionListener(@NonNull OnStartSeleccionListener onStartSeleccionListener) {
		adapter.addOnStartSeleccionListener(onStartSeleccionListener);
	}

	@Override
	public void addOnStopSeleccionListener(@NonNull OnStopSeleccionListener onStopSeleccionListener) {
		adapter.addOnStopSeleccionListener(onStopSeleccionListener);
	}

	@Override
	public void addOnSeleccionChangeListener(@NonNull OnSeleccionChangeListener onSeleccionChangeListener) {
		adapter.addOnSeleccionChangeListener(onSeleccionChangeListener);
	}
}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/seleccionable/AbstractListaSeleccionable.kt
package com.victorlh.android.framework.listas.adapter.seleccionable

import com.victorlh.android.framework.listas.adapter.AbstractLista
import com.victorlh.android.framework.listas.adapter.seleccionable.ListaSeleccionable.*

/**
 * Created by victor on 07/07/2017.
 */
abstract class AbstractListaSeleccionable<T> protected constructor(clazz: Class<T>, tipoSeleccion: ETipoSeleccion = ETipoSeleccion.NONE) : AbstractLista<T>(clazz), ListaSeleccionable<T> {

    private val adapter: ListaSeleccionableAdapter<T> by lazy {
        ListaSeleccionableAdapter(tipoSeleccion, this)
    }

    override val tipoSeleccion: ETipoSeleccion
        get() = adapter.tipoSeleccion

    override fun setSeleccion(items: Array<T>) {
        adapter.setSeleccion(items)
    }

    override fun setSeleccion(items: List<T>) {
        adapter.setSeleccion(items)
    }

    override fun addSeleccion(item: T) {
        adapter.addSeleccion(item)
    }

    override fun removeSeleccion(item: T) {
        adapter.removeSeleccion(item)
    }

    override fun removeAllSeleccion() {
        adapter.removeAllSeleccion()
    }

    override fun haySeleccionados(): Boolean {
        return adapter.haySeleccionados()
    }

    override fun isSeleccionado(item: T): Boolean {
        return adapter.isSeleccionado(item)
    }

    override val itemsSeleccionados: Array<T>
        get() = adapter.itemsSeleccionados

    override fun addOnStartSeleccionListener(onStartSeleccionListener: OnStartSeleccionListener) {
        adapter.addOnStartSeleccionListener(onStartSeleccionListener)
    }

    override fun addOnStopSeleccionListener(onStopSeleccionListener: OnStopSeleccionListener) {
        adapter.addOnStopSeleccionListener(onStopSeleccionListener)
    }

    override fun addOnSeleccionChangeListener(onSeleccionChangeListener: OnSeleccionChangeListener) {
        adapter.addOnSeleccionChangeListener(onSeleccionChangeListener)
    }

    override fun addOnItemDeseleccionadoListener(onItemDeseleccionadoListener: OnItemDeseleccionadoListener<T>) {
        adapter.addOnItemDeseleccionadoListener(onItemDeseleccionadoListener)
    }

    override fun addOnItemSeleccionadoListener(onItemSeleccionadoListener: OnItemSeleccionadoListener<T>) {
        adapter.addOnItemSeleccionadoListener(onItemSeleccionadoListener)
    }

}